src/main/java/io/github/thebusybiscuit/slimefun4/utils/InfiniteBlockGenerator.java

```diff
@@ -0,0 +1,177 @@
+package io.github.thebusybiscuit.slimefun4.utils;
+
+import java.util.function.Predicate;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import org.apache.commons.lang.Validate;
+import org.bukkit.Bukkit;
+import org.bukkit.Material;
+import org.bukkit.block.Block;
+import org.bukkit.block.BlockFace;
+import org.bukkit.block.BlockState;
+import org.bukkit.event.block.BlockFormEvent;
+
+import io.github.thebusybiscuit.slimefun4.api.MinecraftVersion;
+import io.github.thebusybiscuit.slimefun4.implementation.items.androids.MinerAndroid;
+import io.papermc.lib.PaperLib;
+
+/**
+ * This enum holds various ways of infinite block generators.
+ * The most prominent member of these is the standard Cobblestone Generator.
+ * We use this enum for performance optimizations for the {@link MinerAndroid}.
+ * 
+ * @author TheBusyBiscuit
+ *
+ */
+public enum InfiniteBlockGenerator implements Predicate<Block> {
+
+    /**
+     * Your standard Cobblestone Generator with flowing lava and water.
+     */
+    COBBLESTONE_GENERATOR("COBBLESTONE"),
+
+    /**
+     * When lava flows onto a stationary water block it generates normal stone.
+     */
+    STONE_GENERATOR("STONE"),
+
+    /**
+     * The Basalt Generator (1.16+ only) allows you to generate infinite Basalt!
+     */
+    BASALT_GENERATOR("BASALT");
+
+    public static final InfiniteBlockGenerator[] values = values();
+    private static final BlockFace[] sameLevelFaces = { BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH, BlockFace.WEST };
+
+    private final Material material;
+
+    InfiniteBlockGenerator(@Nonnull String type) {
+        this.material = Material.matchMaterial(type);
+    }
+
+    /**
+     * This returns the generated {@link Material} of this {@link InfiniteBlockGenerator}.
+     * This method can return null if the associated {@link Material} is not available in the current
+     * {@link MinecraftVersion}.
+     * 
+     * @return The generated {@link Material} or null
+     */
+    @Nullable
+    public Material getGeneratedMaterial() {
+        return material;
+    }
+
+    /**
+     * Similar to {@link #test(Block)} this tests whether this {@link InfiniteBlockGenerator}
+     * exists at the given {@link Block}.
+     * 
+     * @param b
+     *            The {@link Block}
+     * 
+     * @return Whether this {@link InfiniteBlockGenerator} exists at the given {@link Block}
+     */
+    @Override
+    public boolean test(@Nonnull Block b) {
```

Comment:
naming wise I feel like `InfiniteBlockGenerator#checkForAt(Block b)` would fit better than `InfiniteBlockGenerator#test(Block b)`

Reply:
This method is overridden, as this is a Predicate implementation.